packages/react-components/react-avatar/src/components/AvatarGroup/useAvatarGroup.tsx

```diff
@@ -0,0 +1,94 @@
+import * as React from 'react';
+import { getNativeElementProps, resolveShorthand } from '@fluentui/react-utilities';
+import { MoreHorizontalRegular } from '@fluentui/react-icons';
+import { PopoverSurface } from '@fluentui/react-popover';
+import type { AvatarGroupProps, AvatarGroupState } from './AvatarGroup.types';
+
+/**
+ * Create the state required to render AvatarGroup.
+ *
+ * The returned state can be modified with hooks such as useAvatarGroupStyles_unstable,
+ * before being passed to renderAvatarGroup_unstable.
+ *
+ * @param props - props from this instance of AvatarGroup
+ * @param ref - reference to root HTMLElement of AvatarGroup
+ */
+export const useAvatarGroup_unstable = (props: AvatarGroupProps, ref: React.Ref<HTMLElement>): AvatarGroupState => {
+  const { children, layout = 'spread', maxAvatars = 5, size = 32 } = props;
+  const { overflowIndicator = size < 24 ? 'icon' : 'count' } = props;
+  const childrenArray = React.Children.toArray(children);
+
+  let rootChildren = childrenArray;
+  let overflowChildren;
+  let overflowButtonChildren;
+
+  if (layout === 'pie') {
+    rootChildren = childrenArray.slice(0, 3);
+    overflowChildren = childrenArray;
+  } else if (childrenArray.length > maxAvatars) {
+    const numOfAvatarsToHide = childrenArray.length - maxAvatars + 1;
+
+    rootChildren = childrenArray.slice(numOfAvatarsToHide);
+    overflowChildren = childrenArray.slice(0, numOfAvatarsToHide);
+
+    if (overflowIndicator === 'icon') {
+      overflowButtonChildren = <MoreHorizontalRegular />;
+    } else {
+      overflowButtonChildren = numOfAvatarsToHide > 99 ? '99+' : `+${numOfAvatarsToHide}`;
+    }
+  }
```

Comment:
Children iteration was proven as a pattern that breaks composition. For example:

```tsx
function App() {
  return (
    <AvatarGroup>
      <>
        <AvatarGroupItem />
        <AvatarGroupItem />
      </>
    </AvatarGroup>
  );
}
```

```tsx
function AlwaysPresentAvatars() {
  return (
    <>
      <AvatarGroupItem />
      <AvatarGroupItem />
    </>
  );
}

function App() {
  return (
    <AvatarGroup>
      <AlwaysPresentAvatars />
    </AvatarGroup>
  );
}
```

Both snippets will render the same markup, but will not be properly handled by `AvatarGroup` component. I think that we need to figure out a better pattern to handle these scenarios as in current state can lead to unexpected problems and #23498 does not prevent them.